halo2/src/circuit_builder.rs

```diff
@@ -166,6 +170,54 @@ pub(crate) fn analyzed_to_circuit<T: FieldElement>(
             _ => unimplemented!(),
         }
     }
+
+    assert!(
+        analyzed.public_declarations.len() <= num_rows,
+        "More publics than rows!"
+    );
+
+    // Enforce publics by copy-constraining to cells in the instance column.
+    // For example, if we have the following public declarations:
+    // - public out1 = A(2);
+    // - public out2 = B(0);
+    // This would lead to the corresponding values being copied into the public
+    // column as follows:
+    // | Row | A   | B   | public |
+    // |-----|-----|-----|--------|
+    // |  0  |  0  | *5* |  *2*   |
+    // |  1  |  1  |  6  |  *5*   |
+    // |  2  | *2* |  7  |        |
+    // |  3  |  4  |  8  |        |
```

Comment:
So the syntactic id of the public column is its index in the `public` column, right?

Reply:
The index in the public columm is the index of the public declaration in the PIL file.